File: packages/json/llm_json_stream_dart/test/properties/number_property_test.dart
import 'package:llm_json_stream/llm_json_stream.dart';
import 'package:test/test.dart';

import 'list_property_test.dart';

/// Enable verbose logging to debug test execution
const bool verbose = false;

void main() {
  group('Number Property Tests', () {
    test('simple integer', () async {
      if (verbose) print('\n[TEST] Simple integer');

      final json = '{"age":30}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final ageStream = parser.getNumberProperty("age");

      // For atomic values, stream should emit once
      final emittedValue = await ageStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      // Verify future resolves to the same value
      final finalValue = await ageStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(30));
      expect(finalValue, equals(30));
    });

    test('negative integer', () async {
      if (verbose) print('\n[TEST] Negative integer');

      final json = '{"temperature":-5}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final tempStream = parser.getNumberProperty("temperature");

      final emittedValue = await tempStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await tempStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(-5));
      expect(finalValue, equals(-5));
    });

    test('decimal number', () async {
      if (verbose) print('\n[TEST] Decimal number');

      final json = '{"price":19.99}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final priceStream = parser.getNumberProperty("price");

      final emittedValue = await priceStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await priceStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(19.99));
      expect(finalValue, equals(19.99));
    });

    test('scientific notation', () async {
      if (verbose) print('\n[TEST] Scientific notation');

      final json = '{"large":1.23e10}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final largeStream = parser.getNumberProperty("large");

      final emittedValue = await largeStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await largeStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(1.23e10));
      expect(finalValue, equals(1.23e10));
    });

    test('negative scientific notation', () async {
      if (verbose) print('\n[TEST] Negative scientific notation');

      final json = '{"small":-2.5e-3}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final smallStream = parser.getNumberProperty("small");

      final emittedValue = await smallStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await smallStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(-2.5e-3));
      expect(finalValue, equals(-2.5e-3));
    });

    test('zero', () async {
      if (verbose) print('\n[TEST] Zero');

      final json = '{"count":0}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final countStream = parser.getNumberProperty("count");

      final emittedValue = await countStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await countStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(0));
      expect(finalValue, equals(0));
    });

    test('large integer', () async {
      if (verbose) print('\n[TEST] Large integer');

      final json = '{"big":9223372036854775807}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 5,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final bigStream = parser.getNumberProperty("big");

      final emittedValue = await bigStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await bigStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(9223372036854775807));
      expect(finalValue, equals(9223372036854775807));
    });

    test('nested number access', () async {
      if (verbose) print('\n[TEST] Nested number access');

      final json = '{"user":{"age":25}}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final ageStream = parser.getNumberProperty("user.age");

      final emittedValue = await ageStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await ageStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(25));
      expect(finalValue, equals(25));
    });

    test('deeply nested number access', () async {
      if (verbose) print('\n[TEST] Deeply nested number access');

      final json = '{"data":{"stats":{"count":42}}}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 5,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final countStream = parser.getNumberProperty("data.stats.count");

      final emittedValue = await countStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await countStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(42));
      expect(finalValue, equals(42));
    });

    test('decimal with trailing zeros', () async {
      if (verbose) print('\n[TEST] Decimal with trailing zeros');

      final json = '{"value":3.14000}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final valueStream = parser.getNumberProperty("value");

      final emittedValue = await valueStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await valueStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(3.14));
      expect(finalValue, equals(3.14));
    });

    test('number as num type', () async {
      if (verbose) print('\n[TEST] Number as num type');

      final json = '{"value":123.456}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final valueStream = parser.getNumberProperty("value");

      final emittedValue = await valueStream.stream.first;
      if (verbose) print('[EMITTED] $emittedValue');

      final finalValue = await valueStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(emittedValue, equals(123.456));
      expect(finalValue, equals(123.456));
    });
  });
}

File: packages/json/llm-json-stream-ts/packages/llm-json-stream/test/properties/number_property.test.ts
/**
 * Tests for number property parsing
 * Uses async iterators as the primary streaming interface.
 *
 * NOTE: Numbers are atomic values - they emit ONCE when parsing completes,
 * not incrementally like strings.
 */

import { describe, expect, test } from "@jest/globals";
import { JsonStreamParser } from "../../src/classes/json_stream_parser.js";
import { streamTextInChunks } from "../../src/utilities/stream_text_in_chunks.js";

describe("Number Property Tests", () => {
    test("positive integer", async () => {
        const json = '{"age":30}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 4,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const ageStream = parser.getNumberProperty("age");

        const streamEvents: number[] = [];
        for await (const value of ageStream) {
            streamEvents.push(value);
        }

        const finalValue = await ageStream.promise;

        // Number emits once when complete
        expect(streamEvents).toEqual([30]);
        expect(finalValue).toBe(30);
    });

    test("negative integer", async () => {
        const json = '{"temperature":-5}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 6,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const tempStream = parser.getNumberProperty("temperature");

        const finalValue = await tempStream.promise;
        expect(finalValue).toBe(-5);
    });

    test("floating point number", async () => {
        const json = '{"price":19.99}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 5,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const priceStream = parser.getNumberProperty("price");

        const finalValue = await priceStream.promise;
        expect(finalValue).toBe(19.99);
    });

    test("zero", async () => {
        const json = '{"count":0}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 4,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const countStream = parser.getNumberProperty("count");

        const finalValue = await countStream.promise;
        expect(finalValue).toBe(0);
    });

    test("negative zero", async () => {
        const json = '{"value":-0}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 4,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const valueStream = parser.getNumberProperty("value");

        const finalValue = await valueStream.promise;
        expect(finalValue).toBe(-0);
    });

    test("scientific notation", async () => {
        const json = '{"distance":1.5e10}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 6,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const distanceStream = parser.getNumberProperty("distance");

        const finalValue = await distanceStream.promise;
        expect(finalValue).toBe(1.5e10);
    });

    test("large number", async () => {
        const json = '{"population":7800000000}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 8,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const popStream = parser.getNumberProperty("population");

        const finalValue = await popStream.promise;
        expect(finalValue).toBe(7800000000);
    });

    test("number in nested object", async () => {
        const json = '{"user":{"age":25}}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 6,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const ageStream = parser.getNumberProperty("user.age");

        const finalValue = await ageStream.promise;
        expect(finalValue).toBe(25);
    });

    test("number in array", async () => {
        const json = '{"scores":[95,87,92]}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 7,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const score0 = parser.getNumberProperty("scores[0]");
        const score1 = parser.getNumberProperty("scores[1]");
        const score2 = parser.getNumberProperty("scores[2]");

        const [val0, val1, val2] = await Promise.all([
            score0.promise,
            score1.promise,
            score2.promise,
        ]);

        expect(val0).toBe(95);
        expect(val1).toBe(87);
        expect(val2).toBe(92);
    });

    test("async iterator emits number value once", async () => {
        const json = '{"value":42}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 3,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const valueStream = parser.getNumberProperty("value");

        // Use async iterator pattern
        const values: number[] = [];
        for await (const value of valueStream) {
            values.push(value);
        }

        // Number emits exactly once with the final value
        expect(values.length).toBe(1);
        expect(values[0]).toBe(42);
    });
});

